### PR TITLE
Cleaning up changes and adding Cobra Codex Alterations

### PR DESCRIPTION
**Pull request by JFranklin3000**
##### In this PR
- Cobra Codex Alteration Items Added
- Check to disallow adding the same alteration twice
- originalId for almost all items dropped on an actor

##### Testing
- Validate new Items
- drop prevention for too many movement reductions, duplicates, etc. 
- items are in ccitems

### CODE REVIEW

File: template.json [hunk 1]
```diff
@@ -642,12 +642,18 @@
           "book": "",
           "page": null
         }
+      },
+      "item": {
```
Comment (phildominguez) [inline template.json:646]: I think this name is too vague for me. What about `uniqueItem`? Or maybe just `originalItem`.

File: module/helpers/utils.mjs [hunk 2]
```diff
@@ -70,6 +70,24 @@ function _localizeObject(obj, keys) {
   }
 }
 
+export function parseId (uuid) {
```
Comment (phildominguez) [inline module/helpers/utils.mjs:73]:
```suggestion
export function parseId(uuid) {
```
Don't forget method desc block

File: templates/item/sheets/alteration.hbs [hunk 3]
```diff
@@ -12,81 +12,116 @@
...
+          {{/each}}
+          {{/select}}
+        </select>
+        {{/inline}}
+        {{/"systems/essence20/templates/item/parts/sheet-field.hbs"}}
+      {{/ifEquals}}
+
+      {{#ifEquals system.alterationType 'other'}}
+
+        {{!-- Alteration Non Essence Bonus --}}
```
Comment (phildominguez) [inline templates/item/sheets/alteration.hbs:111]:
```suggestion
      {{#ifEquals system.alterationType 'other'}}
        {{!-- Alteration Non Essence Bonus --}}
```

File: templates/item/sheets/alteration.hbs [hunk 4]
```diff
@@ -12,81 +12,116 @@
...
       {{/inline}}
       {{/"systems/essence20/templates/item/parts/sheet-field.hbs"}}
 
-      {{!-- Alteration Non Essence Bonus --}}
-      {{#> "systems/essence20/templates/item/parts/sheet-field.hbs" label='E20.Bonus'}}
-      {{#*inline "item-field-inputs"}}
-      <input type="number" name="system.bonus" value="{{system.bonus}}" />
-      {{/inline}}
-      {{/"systems/essence20/templates/item/parts/sheet-field.hbs"}}
+      {{#ifEquals system.alterationType 'essence'}}
```
Comment (phildominguez) [inline templates/item/sheets/alteration.hbs:43]:
```suggestion
      {{#ifEquals system.alterationType 'essence'}}
```

**Review by phildominguez — CHANGES_REQUESTED** [review]

File: module/sheet-handlers/alteration-handler.mjs [hunk 5]
```diff
@@ -22,10 +23,21 @@ export class AlterationHandler {
   * @param {Function} dropFunc   The function to call to complete the Alteration drop
   */
   async alterationUpdate(alteration, dropFunc) {
-    if (alteration.system.essenceBonus) {
-      await this._showAlterationBonusSkillDialog(alteration, dropFunc);
-    } else if (alteration.system.bonusMovement) {
-      await this._showAlterationCostMovementDialog(alteration, dropFunc);
+    const alterationUuid = parseId(alteration.uuid);
+    console.log(alterationUuid);
+    for (let actorItem of this._actor.items) {
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply module/sheet-handlers/alteration-handler.mjs:28]: both fixed

File: module/sheet-handlers/alteration-handler.mjs [hunk 6]
```diff
@@ -22,10 +23,21 @@ export class AlterationHandler {
...
+          ui.notifications.warn(game.i18n.localize('E20.AlterationAlreadyTaken'));
+          return;
+        }
+      }
+    }
+
+    if (alteration.system.alterationType == 'essence') {
+      await this._showAlterationBonusSkillDialog(alteration, dropFunc, alterationUuid);
+    } else if (alteration.system.alterationType == 'movement') {
+      await this._showAlterationCostMovementDialog(alteration, dropFunc, alterationUuid);
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply module/sheet-handlers/alteration-handler.mjs]: all fixed

File: module/sheet-handlers/alteration-handler.mjs [hunk 7]
```diff
@@ -123,7 +142,7 @@ export class AlterationHandler {
   * @param {Alteration} alteration The alteration
   * @param {Function} dropFunc   The function to call to complete the Alteration drop
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply module/sheet-handlers/alteration-handler.mjs]: fixed

File: module/sheet-handlers/alteration-handler.mjs [hunk 8]
```diff
@@ -34,7 +46,12 @@ export class AlterationHandler {
...
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply module/sheet-handlers/alteration-handler.mjs]: fixed

File: module/sheet-handlers/alteration-handler.mjs [hunk 9]
```diff
@@ -34,7 +46,12 @@ export class AlterationHandler {
...
   */
-  async _showAlterationCostMovementDialog (alteration, dropFunc) {
+  async _showAlterationCostMovementDialog (alteration, dropFunc, alterationUuid) {
+    if (!this._actor.system.movement.ground.base && !this._actor.system.movement.aerial.base && !this._actor.system.movement.climb.base && !this._actor.system.movement.swim.base) {
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply module/sheet-handlers/alteration-handler.mjs]: fixed

File: module/sheet-handlers/alteration-handler.mjs [hunk 10]
```diff
@@ -159,7 +178,7 @@ export class AlterationHandler {
         buttons: {
           save: {
             label: game.i18n.localize('E20.AcceptButton'),
-            callback: html => this._processAlterationSkillIncrease(alteration, rememberOptions(html), dropFunc),
+            callback: html => this._processAlterationSkillIncrease(alteration, rememberOptions(html), dropFunc, alterationUuid),
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply module/sheet-handlers/alteration-handler.mjs]: fixed

File: module/helpers/utils.mjs (hunk 2 above)
Comment (JFranklin3000) [reply module/helpers/utils.mjs]: fixed

File: templates/item/sheets/alteration.hbs (hunk 3 above)
Comment (JFranklin3000) [reply templates/item/sheets/alteration.hbs]: fixed

File: templates/item/sheets/alteration.hbs (hunk 4 above)
Comment (JFranklin3000) [reply templates/item/sheets/alteration.hbs]: fixed

File: template.json (hunk 1 above)
Comment (JFranklin3000) [reply template.json:646]: I was thinking of this as the template that goes on all items.

File: template.json [hunk 11]
```diff
@@ -642,12 +642,18 @@
...
+        "originalId": ""
       }
     },
     "alteration": {
       "templates": [
-        "itemDescription"
+        "itemDescription",
+        "item"
       ],
+      "alterationType": "other",
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply template.json:655]: I was thinking of it as a way to differentiate from item type which is already there.

File: module/helpers/utils.mjs [hunk 12]
```diff
@@ -70,7 +70,11 @@ function _localizeObject(obj, keys) {
   }
 }
 
-export function parseId (uuid) {
+/*
+* Parse the UUID to get just the ID value of the item
+* @param {uuid} uuid of the item that we are parsing for the id
+*/
+export function parseId(uuid) {
```
Comment (phildominguez) [inline module/helpers/utils.mjs:77]: Return type is String or null for desc

File: module/sheet-handlers/alteration-handler.mjs [hunk 13]
```diff
@@ -93,9 +98,10 @@ export class AlterationHandler {
   * Handles the movements choices and updating the actors movements
   * @param {Alteration} alteration The alteration
   * @param {Options} options  The options selected from the dialog
+  * @param {AlterationUUID} alterationUuid The original ID of the alteration
```
Comment (phildominguez) [inline module/sheet-handlers/alteration-handler.mjs:101]: These are just String

File: template.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (phildominguez) [reply template.json:655]: Generally in the template we've just used `type` though, like Gear, Perk, Upgrade

File: template.json (hunk 1 above)
Comment (phildominguez) [reply template.json:646]: That works, just move it before `itemDescription` everywhere to keep it alphabetical.

File: module/helpers/utils.mjs [hunk 14]
```diff
@@ -70,6 +70,28 @@ function _localizeObject(obj, keys) {
   }
 }
 
+/*
+* Parse the UUID to get just the ID value of the item
+* @param {string} uuid of the item that we are parsing for the id
+* @return {string} index or null returned.
```
Comment (phildominguez) [inline module/helpers/utils.mjs:76]:
```suggestion
* @return {string|null} index or null returned.
```

File: module/helpers/utils.mjs [hunk 15]
```diff
@@ -70,6 +70,28 @@ function _localizeObject(obj, keys) {
...
+    const [, id] = parts;
+    index = id;
+
+  }
+
+  return index || null;
+}
+
+
 /**
```
Comment (phildominguez) [inline module/helpers/utils.mjs:95]:
```suggestion
    index = id;
  }

  return index || null;
}

/**
```

File: module/sheet-handlers/alteration-handler.mjs [hunk 16]
```diff
@@ -17,24 +18,45 @@ export class AlterationHandler {
...
+      await this._showAlterationBonusSkillDialog(alteration, alterationUuid, dropFunc);
+    } else if (alteration.system.type == 'movement') {
+      await this._showAlterationCostMovementDialog(alteration, alterationUuid, dropFunc);
     }
   }
 
   /**
   * Handles the creation of a dialog to create alteration movement reduction
   * @param {Alteration} alteration The alteration
+  * @param {AlterationUUID} alterationUuid The original ID of the alteration
```
Comment (phildominguez) [inline module/sheet-handlers/alteration-handler.mjs:47]:
```suggestion
  * @param {String} alterationUuid The original ID of the alteration
```

File: module/sheet-handlers/alteration-handler.mjs [hunk 17]
```diff
@@ -76,11 +98,10 @@ export class AlterationHandler {
   * Handles the movements choices and updating the actors movements
   * @param {Alteration} alteration The alteration
   * @param {Options} options  The options selected from the dialog
+  * @param {string} alterationUuid The original ID of the alteration
```
Comment (phildominguez) [inline module/sheet-handlers/alteration-handler.mjs:101]:
```suggestion
  * @param {String} alterationUuid The original ID of the alteration
```
We sort of mix caps between different files, but we should try to be consistent in the same file

File: module/sheet-handlers/alteration-handler.mjs [hunk 18]
```diff
@@ -115,15 +139,17 @@ export class AlterationHandler {
...
     await newAlteration.update ({
       "system.movementCost": options,
+      "system.originalId": alterationUuid,
     });
   }
 
   /**
   * Handles creating a list to select a skill to increase
   * @param {Alteration} alteration The alteration
+  * @param {AlterationUUID} alterationUuid The original ID of the alteration
```
Comment (phildominguez) [inline module/sheet-handlers/alteration-handler.mjs:149]:
```suggestion
  * @param {String} alterationUuid The original ID of the alteration
```

File: module/sheet-handlers/alteration-handler.mjs [hunk 19]
```diff
@@ -170,9 +196,10 @@ export class AlterationHandler {
   * Handles choosing a skill to increase
   * @param {Alteration} alteration The alteration
   * @param {Object} options    The options resulting from _showAlterationBonusSkillDialog()
+  * @param {AlterationUUID} alterationUuid The original ID of the alteration
```
Comment (phildominguez) [inline module/sheet-handlers/alteration-handler.mjs:199]:
```suggestion
  * @param {String} alterationUuid The original ID of the alteration
```

File: module/sheet-handlers/alteration-handler.mjs [hunk 20]
```diff
@@ -187,19 +214,20 @@ export class AlterationHandler {
...
-      await this._showAlterationCostSkillDialog(alteration, bonusSkill, dropFunc);
+      await this._showAlterationCostSkillDialog(alteration, bonusSkill, alterationUuid, dropFunc);
     }
   }
 
   /**
   * Handles creating a dialog to choose an essencd to decrease
   * @param {Object} alteration The alteration
   * @param {String} bonusSkill    The skill selected from _processAlterationSkillIncrease()
+  * @param {AlterationUUID} alterationUuid The original ID of the alteration
```
Comment (phildominguez) [inline module/sheet-handlers/alteration-handler.mjs:227]:
```suggestion
  * @param {String} alterationUuid The original ID of the alteration
```

File: module/sheet-handlers/alteration-handler.mjs [hunk 21]
```diff
@@ -228,10 +256,11 @@ export class AlterationHandler {
   * Handles creating a dialog to select a skill to decrease
   * @param {Alteration} alteration The alteration
   * @param {String} bonusSkill    The skill selected from _processAlterationSkillIncrease()
-  * @param {Function} dropFunc   The function to call to complete the Alteration drop
   * @param {Object} options    The options resulting from _showAlterationCostEssenceDialog()
+  * @param {AlterationUUID} alterationUuid The original ID of the alteration
```
Comment (phildominguez) [inline module/sheet-handlers/alteration-handler.mjs:260]:
```suggestion
  * @param {String} alterationUuid The original ID of the alteration
```

File: module/sheet-handlers/alteration-handler.mjs [hunk 22]
```diff
@@ -312,11 +341,12 @@ export class AlterationHandler {
   * Handles the setting of values from a essence bonus alteration
   * @param {Alteration} alteration The alteration
   * @param {String} bonusSkill    The skill selected from _processAlterationSkillIncrease()
-  * @param {Function} dropFunc   The function to call to complete the Alteration drop
   * @param {String} costEssence    The essence selected from _showAlterationCostEssenceDialog()
   * @param {Object} options    The options resulting from _showAlterationCostSkillDialog()
+  * @param {AlterationUUID} alterationUuid The original ID of the alteration
```
Comment (phildominguez) [inline module/sheet-handlers/alteration-handler.mjs:346]:
```suggestion
  * @param {String} alterationUuid The original ID of the alteration
```

File: module/sheets/actor-sheet.mjs [hunk 23]
```diff
@@ -585,8 +589,18 @@ export class Essence20ActorSheet extends ActorSheet {
...
+      itemUuid = await parseId(data.uuid);
+
+      droppedItemList = await super._onDropItem(event, data);
+      newItem = droppedItemList[0];
+
+      await newItem.update ({
+        "system.originalId": itemUuid,
+      });
+
+      return droppedItemList;
```
Comment (phildominguez) [inline module/sheets/actor-sheet.mjs:603]: Just thought of this, what about making a separate function like `_onDropDefault` that does all this stuff, then we don't need to declare the 3 vars above and it'll be a bit nicer.

File: module/sheets/actor-sheet.mjs [hunk 24]
```diff
@@ -627,6 +641,7 @@ export class Essence20ActorSheet extends ActorSheet {
     return weaponList;
   }
 
+
   /**
```
Comment (phildominguez) [inline module/sheets/actor-sheet.mjs:645]:
```suggestion
  }

  /**
```

**Review by phildominguez — CHANGES_REQUESTED** [review]

File: module/sheets/actor-sheet.mjs (hunk 23 above)
Comment (JFranklin3000) [reply module/sheets/actor-sheet.mjs]: changed

File: module/sheet-handlers/alteration-handler.mjs [hunk 25]
```diff
@@ -260,7 +260,7 @@ export class AlterationHandler {
   * @param {String} alterationUuid The original ID of the alteration
   * @param {Function} dropFunc   The function to call to complete the Alteration drop
   */
-  async _showAlterationCostSkillDialog(alteration, bonusSkill, options, alterationUuid, dropFunc) {
+  async _showAlterationCostSkillDialog(alteration, bonusSkill, alterationUuid, dropFunc, options) {
```
Comment (phildominguez) [inline module/sheet-handlers/alteration-handler.mjs:263]: I would keep dropFunc at the end like elsewhere

**Review by phildominguez — CHANGES_REQUESTED** [review]

File: module/sheet-handlers/alteration-handler.mjs (hunk 25 above)
Comment (JFranklin3000) [reply module/sheet-handlers/alteration-handler.mjs:263]: If I do that then we will bring back the issue you found earlier when options is empty.

File: lang/en.json [hunk 26]
```diff
@@ -79,13 +79,25 @@
     "ActorSizeTowering": "Towering",
     "ActorStun": "Stun",
     "Alteration": "Alteration",
+    "AlterationAlreadyTaken": "You have already add this alteration",
```
Comment (phildominguez) [inline lang/en.json:82]:
```suggestion
    "AlterationAlreadyTaken": "You have already added this alteration",
```

**Review by phildominguez — CHANGES_REQUESTED** [review]

**Review by phildominguez — APPROVED** [review]

**Pull request merged**